client/src/portal/mod.rs

```diff
@@ -375,8 +375,21 @@ impl Keyring {
 
     /// Return key, derive and store it first if not initialized
     async fn derive_key(&self) -> Arc<Key> {
+        let keyring = Arc::clone(&self.keyring);
+        let secret = Arc::clone(&self.secret);
+
         if self.key.lock().await.is_none() {
```

Comment:
the guard is not dropped here, are you sure this solves the blocking issue? you should probably test key rack with this branch of oo7